### DISCUSSION TITLE
Ordering of slider elements for legacy skins

### DISCUSSION THREAD

**Discussion by LittleEndu** [ppy/osu#14594] [Q&A]
Opening a discussion to list things I have tried to fix the ordering of how slider elements are drawn for legacy skins.

The ordering should be changed from the left to the right:
```
slider start circle overlay.png		sliderb.png
slider start circle.png        		slider start circle overlay.png              		
sliderb.png                    		reverse arrow.png
reverse arrow.png              		slider start circle.png
slider end circle overlay.png  		slider end circle overlay.png
slider end circle.png          		slider end circle.png
```

Doing so is not trivial without introducing needless legacy logic to `DrawableSlider` that could make it difficult to implement the new skinning system. 

I've tried making a new container in `LegacyMainCirclePiece` that could accept the `reverse arrow.png` element from `ReverseArrowPiece` (or its new legacy counterpart) as a proxy, but that runs into an issue where I assume the `SliderHeadCircle` reaches the end of its lifetime and the `DrawableSliderHeadCircle` dies with it.

Instead of moving the `reverse arrow.png` above `slider start circle.png` I have tried moving the `slider start circle.png` down. `DrawableSlider` has `tickContainer` that if we could access then  `LegacyMainCirclePiece` could proxy its `hitCircleSprite` into it, as the `tickContainer` is below the `reverse arrow.png`, but sadly it would also be below `DrawableSliderRepeat`'s `slider end circle overlay.png` and `slider end circle.png`.

Because the only way to achieve different behaviour default skin vs legacy skin is to use `SkinnableDrawable`, I think `DrawableSlider` instead of being a component for `Container`s needs to just have one `SkinnableDrawable` or some new subclass of it that deals with ordering slider elements. Then the `DrawableSlider` instead of using its own `Container`s would add nested hitobjects to that new `SkinnableDrawable` subclass instead.

**Comment by peppy** [comment]
Why is `sliderb-spec.png` about the follow circle? That sounds objectively wrong. Is this used by any existing skins?

**Reply by LittleEndu** [reply]
Maybe copy-paste oversight on how it is in lazer, but the right side (how it is on stable) for the slider ball is copied from bdach's investigation into stable code

**Reply by peppy** [reply]
I don't care how the code is written though. We have never been looking for 1:1 code reproduction. We're trying to attain user expected behaviour with the *minimal* hackiness/incorrectness possible.

Maybe you should rewrite this discussion OP to mention the pieces which are actually not meeting user expectations right now. ie. is there anything other than the repeat arrow which is currently an issue?

**Reply by LittleEndu** [reply]
Didn't want to leave out any info but I guess I did include too much. 

There shouldn't be any other issues than the reverse arrow. 

(I'll edit the OP later when I'm at home)

**Comment by Walavouchey** [comment]
the pros of the stable ordering essentially boil down to ease of readability. overlapping hit objects are a common occurrence in maps

below is a 2b map to illustrate that in stable (osu! game mode), slider bodies are always drawn below slider heads, slider tails and hit circles, basically on a separate layer.

lazer with argon, classic, my skin, and triangles:

https://user-images.githubusercontent.com/36758269/216821377-8d61f57b-df14-4ac5-b328-64638272b91d.mp4

stable with default skin:

https://user-images.githubusercontent.com/36758269/216821919-f771bd85-11f6-485b-92e4-c7255ee65484.mp4

copying what i've said in https://github.com/ppy/osu/discussions/21994 to what i think is a more appropriate spot (and to hopefully visuals illustrate the superiority of stable ordering better)

**Reply by AndersGitTemp** [reply]
Sorry to revive an old topic, but has there been any follow up to this? Do we know if it will be fixed at some point in Lazer so it won't butcher a map like this one?

**Reply by peppy** [reply]
Nothing yet. It's a hard one as it breaks everything we have setup to date. It also breaks visual expectations.
